File: examples/typescript/src/merkle-leaves-list.ts
// Merkle leaves-list CBOR codec for CIP-309.
// Spec: CIP-309 §6.5 (canonical CBOR wire form),
//       CIP-309 §4.5 (on-chain commitment).
//
// CBOR is the normative wire form of the leaves-list file published to the
// content-addressed substrate referenced by `merkle[i].uris[]`. The JSON
// projection emitted by `toJsonProjection` is informative only: it is a
// human-readable companion view (CLI dumps, doc examples) and MUST NOT be
// used as the byte-normative form when both forms are present.
//
// Final CDDL (CIP-309 §6.5):
//
//   leaves-list = {
//     "format":     "cardano-poe-merkle-leaves-v1",
//     "tree_alg":   "rfc9162-sha256",
//     "root":       bytes .size 32,
//     "leaves":     [ + bytes .size 32 ],
//     "leaf_count": uint,
//     ? "leaf_alg": tstr,
//   }
//
// Unknown / mismatched `format` values throw `SCHEMA_MERKLE_LEAVES_FORMAT_UNSUPPORTED`
// (CIP-309).

import { decodeCanonicalCbor, encodeCanonicalCbor } from './cbor-canonical.ts';

// === Constants ===

const LEAVES_LIST_FORMAT = 'cardano-poe-merkle-leaves-v1';
const TREE_ALG_RFC9162 = 'rfc9162-sha256';
const DIGEST_LENGTH = 32;
const REGISTERED_FORMATS = new Set<string>([LEAVES_LIST_FORMAT]);

// === Types ===

export interface DecodedLeavesList {
  readonly format: typeof LEAVES_LIST_FORMAT;
  readonly treeAlg: 'rfc9162-sha256';
  readonly root: Uint8Array;
  readonly leaves: Uint8Array[];
  readonly leafCount: number;
  readonly leafAlg?: string;
}

export interface EncodeLeavesListArgs {
  root: Uint8Array;
  leaves: Uint8Array[];
  leafCount: number;
  leafAlg?: string;
}

export type LeavesListErrorCode =
  | 'SCHEMA_MERKLE_LEAVES_FORMAT_UNSUPPORTED'
  | 'SCHEMA_MERKLE_LEAVES_MALFORMED';

export class LeavesListError extends Error {
  readonly code: LeavesListErrorCode;
  constructor(code: LeavesListErrorCode, message?: string) {
    super(message ? `${code}: ${message}` : code);
    this.code = code;
    this.name = 'LeavesListError';
  }
}

// === Encode ===

/**
 * Emit canonical CBOR bytes for a leaves-list. The returned bytes are byte-
 * stable across producers per RFC 8949 §4.2.1.
 *
 * Map-key order is implied by `cbor-canonical.ts`; the keys sort as:
 *   `root` (4B) < `format` (6B) < `leaves` (6B) < `leaf_alg` (8B)
 *   < `tree_alg` (8B) < `leaf_count` (10B).
 */
export function encodeLeavesList(args: EncodeLeavesListArgs): Uint8Array {
  if (args.root.length !== DIGEST_LENGTH) {
    throw new LeavesListError(
      'SCHEMA_MERKLE_LEAVES_MALFORMED',
      `root length ${args.root.length} != ${DIGEST_LENGTH}`,
    );
  }
  if (args.leaves.length < 1) {
    throw new LeavesListError(
      'SCHEMA_MERKLE_LEAVES_MALFORMED',
      'leaves array MUST be non-empty',
    );
  }
  for (let i = 0; i < args.leaves.length; i++) {
    const leaf = args.leaves[i]!;
    if (!(leaf instanceof Uint8Array) || leaf.length !== DIGEST_LENGTH) {
      throw new LeavesListError(
        'SCHEMA_MERKLE_LEAVES_MALFORMED',
        `leaves[${i}] MUST be a Uint8Array(${DIGEST_LENGTH})`,
      );
    }
  }
  if (!Number.isInteger(args.leafCount) || args.leafCount < 1) {
    throw new LeavesListError(
      'SCHEMA_MERKLE_LEAVES_MALFORMED',
      `leaf_count must be a positive integer; got ${String(args.leafCount)}`,
    );
  }
  const map: Record<string, unknown> = {
    format: LEAVES_LIST_FORMAT,
    tree_alg: TREE_ALG_RFC9162,
    root: args.root,
    leaves: args.leaves,
    leaf_count: args.leafCount,
  };
  if (args.leafAlg !== undefined) {
    map['leaf_alg'] = args.leafAlg;
  }
  return encodeCanonicalCbor(map);
}

// === Decode ===

/**
 * Parse canonical CBOR bytes as a leaves-list and validate the schema. The
 * `format` field is the version hook (CIP-309 §6.5); a value not in the
 * registered set raises `SCHEMA_MERKLE_LEAVES_FORMAT_UNSUPPORTED`
 * (CIP-309).
 */
export function decodeLeavesList(bytes: Uint8Array): DecodedLeavesList {
  const decoded = decodeCanonicalCbor(bytes);
  if (typeof decoded !== 'object' || decoded === null || Array.isArray(decoded)) {
    throw new LeavesListError(
      'SCHEMA_MERKLE_LEAVES_MALFORMED',
      'top-level leaves-list MUST be a CBOR map',
    );
  }
  const m = decoded as Record<string, unknown>;
  const format = m['format'];
  if (typeof format !== 'string') {
    throw new LeavesListError(
      'SCHEMA_MERKLE_LEAVES_MALFORMED',
      'format MUST be a tstr',
    );
  }
  if (!REGISTERED_FORMATS.has(format)) {
    throw new LeavesListError(
      'SCHEMA_MERKLE_LEAVES_FORMAT_UNSUPPORTED',
      `format '${format}' is not in the registered set`,
    );
  }
  const treeAlg = m['tree_alg'];
  if (treeAlg !== TREE_ALG_RFC9162) {
    throw new LeavesListError(
      'SCHEMA_MERKLE_LEAVES_MALFORMED',
      `tree_alg '${String(treeAlg)}' is not '${TREE_ALG_RFC9162}'`,
    );
  }
  const root = m['root'];
  if (!(root instanceof Uint8Array) || root.length !== DIGEST_LENGTH) {
    throw new LeavesListError(
      'SCHEMA_MERKLE_LEAVES_MALFORMED',
      'root MUST be a 32-byte bstr',
    );
  }
  const leavesRaw = m['leaves'];
  if (!Array.isArray(leavesRaw) || leavesRaw.length < 1) {
    throw new LeavesListError(
      'SCHEMA_MERKLE_LEAVES_MALFORMED',
      'leaves MUST be a non-empty array',
    );
  }
  const leaves: Uint8Array[] = leavesRaw.map((leaf, i) => {
    if (!(leaf instanceof Uint8Array) || leaf.length !== DIGEST_LENGTH) {
      throw new LeavesListError(
        'SCHEMA_MERKLE_LEAVES_MALFORMED',
        `leaves[${i}] MUST be a 32-byte bstr`,
      );
    }
    return leaf;
  });
  const leafCount = m['leaf_count'];
  if (typeof leafCount !== 'number' || !Number.isInteger(leafCount) || leafCount < 1) {
    throw new LeavesListError(
      'SCHEMA_MERKLE_LEAVES_MALFORMED',
      'leaf_count MUST be a positive CBOR uint',
    );
  }
  const out: DecodedLeavesList = {
    format: LEAVES_LIST_FORMAT,
    treeAlg: TREE_ALG_RFC9162,
    root,
    leaves,
    leafCount,
    ...(typeof m['leaf_alg'] === 'string' ? { leafAlg: m['leaf_alg'] as string } : {}),
  };
  return out;
}

// === JSON projection (informative; NOT for verification) ===

/**
 * Render a decoded leaves-list as a JCS-canonicalised JSON string for
 * inspection (CLI dumps, doc examples, debugging). The output is human-
 * readable and stable for byte-level diffing but is NOT the wire form.
 * A verifier MUST NOT use the JSON projection as the byte-normative
 * leaves-list (CIP-309 §6.5).
 */
export function toJsonProjection(decoded: DecodedLeavesList): string {
  const obj: Record<string, unknown> = {
    format: decoded.format,
    tree_alg: decoded.treeAlg,
    root: bytesToHex(decoded.root),
    leaves: decoded.leaves.map(bytesToHex),
    leaf_count: decoded.leafCount,
  };
  if (decoded.leafAlg !== undefined) {
    obj['leaf_alg'] = decoded.leafAlg;
  }
  // JCS (RFC 8785) requires sorted keys + canonical number forms. The keys
  // here are alphabetised by construction.
  const sorted = sortKeysRecursive(obj);
  return JSON.stringify(sorted);
}

/**
 * Inverse of `toJsonProjection`. Provided for completeness; NOT for
 * verification: callers using this function in a verification context
 * MUST treat the result as informative and emit
 * `MERKLE_LEAVES_INFORMATIVE_FORM` (info-severity) per CIP-309.
 */
export function fromJsonProjection(json: string): DecodedLeavesList {
  // eslint-disable-next-line no-console
  if (typeof process !== 'undefined' && process.env?.['CIP309_VERIFICATION_CTX']) {
    console.warn(
      'merkle-leaves-list: fromJsonProjection used in a verification context; ' +
        'CBOR is the normative wire form per CIP-309 §6.5. Emit MERKLE_LEAVES_INFORMATIVE_FORM.',
    );
  }
  const parsed = JSON.parse(json) as Record<string, unknown>;
  const format = parsed['format'];
  if (typeof format !== 'string' || !REGISTERED_FORMATS.has(format)) {
    throw new LeavesListError(
      'SCHEMA_MERKLE_LEAVES_FORMAT_UNSUPPORTED',
      `format '${String(format)}' is not in the registered set`,
    );
  }
  const treeAlg = parsed['tree_alg'];
  if (treeAlg !== TREE_ALG_RFC9162) {
    throw new LeavesListError(
      'SCHEMA_MERKLE_LEAVES_MALFORMED',
      `tree_alg '${String(treeAlg)}' is not '${TREE_ALG_RFC9162}'`,
    );
  }
  const rootHex = parsed['root'];
  if (typeof rootHex !== 'string') {
    throw new LeavesListError(
      'SCHEMA_MERKLE_LEAVES_MALFORMED',
      'root MUST be a hex string',
    );
  }
  const root = hexToBytes(rootHex);
  const leavesRaw = parsed['leaves'];
  if (!Array.isArray(leavesRaw) || leavesRaw.length < 1) {
    throw new LeavesListError(
      'SCHEMA_MERKLE_LEAVES_MALFORMED',
      'leaves MUST be a non-empty array of hex strings',
    );
  }
  const leaves = leavesRaw.map((h, i) => {
    if (typeof h !== 'string') {
      throw new LeavesListError(
        'SCHEMA_MERKLE_LEAVES_MALFORMED',
        `leaves[${i}] MUST be a hex string`,
      );
    }
    return hexToBytes(h);
  });
  const leafCount = parsed['leaf_count'];
  if (typeof leafCount !== 'number' || !Number.isInteger(leafCount) || leafCount < 1) {
    throw new LeavesListError(
      'SCHEMA_MERKLE_LEAVES_MALFORMED',
      'leaf_count MUST be a positive integer',
    );
  }
  const out: DecodedLeavesList = {
    format: LEAVES_LIST_FORMAT,
    treeAlg: TREE_ALG_RFC9162,
    root,
    leaves,
    leafCount,
    ...(typeof parsed['leaf_alg'] === 'string'
      ? { leafAlg: parsed['leaf_alg'] as string }
      : {}),
  };
  return out;
}

// === Helpers ===

function bytesToHex(b: Uint8Array): string {
  let s = '';
  for (let i = 0; i < b.length; i++) s += b[i]!.toString(16).padStart(2, '0');
  return s;
}

function hexToBytes(hex: string): Uint8Array {
  const clean = hex.startsWith('0x') ? hex.slice(2) : hex;
  if (clean.length % 2 !== 0) {
    throw new LeavesListError(
      'SCHEMA_MERKLE_LEAVES_MALFORMED',
      'hex string has odd length',
    );
  }
  const out = new Uint8Array(clean.length / 2);
  for (let i = 0; i < out.length; i++) {
    out[i] = parseInt(clean.slice(i * 2, i * 2 + 2), 16);
  }
  return out;
}

function sortKeysRecursive(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map(sortKeysRecursive);
  }
  if (value !== null && typeof value === 'object') {
    const out: Record<string, unknown> = {};
    const keys = Object.keys(value as Record<string, unknown>).sort();
    for (const k of keys) {
      out[k] = sortKeysRecursive((value as Record<string, unknown>)[k]);
    }
    return out;
  }
  return value;
}

File: examples/python/cip309_examples/merkle_leaves_list.py
"""CIP-309 v1 reference implementation — Merkle leaves-list codec.

The on-storage byte-normative form of the leaves-list file is **canonical
CBOR** (RFC 8949 §4.2). Producers publish CBOR bytes to the content-addressed
substrate referenced by `merkle[i].uris[]`; verifiers parse CBOR. A JSON+JCS
projection is a permitted informative companion (CLI dump, documentation
example) but is NOT the wire form and MUST NOT be treated as authoritative if
both forms are present and they disagree byte-for-byte.

Schema (canonical CBOR):

    leaves-list = {
        "format":     "cardano-poe-merkle-leaves-v1",
        "tree_alg":   "rfc9162-sha256",
        "root":       bytes .size 32,
        "leaves":     [ + bytes .size 32 ],
        "leaf_count": uint,
        ? "leaf_alg": tstr,
    }

The JSON projection encodes `root` and every `leaves[i]` as 64-character
lowercase hex strings; the JSON form is JCS-canonicalised (RFC 8785) when
material to byte-stability for inspection-only consumers.

Cross-language parity: byte-identical with the TypeScript reference
(merkle-leaves-list.ts) when fed the same inputs.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any

from .cbor_canonical import decode_canonical_cbor, encode_canonical_cbor

# === Registered format identifiers ============================================

# The literal `format` value bound to the leaves-list CDDL. Future schema
# revisions bump the suffix (`-v2`, etc.) and add the new value to this set;
# v1 verifiers MUST reject any other value with
# `SCHEMA_MERKLE_LEAVES_FORMAT_UNSUPPORTED`.
LEAVES_LIST_FORMAT_V1: str = "cardano-poe-merkle-leaves-v1"
REGISTERED_FORMATS: frozenset[str] = frozenset({LEAVES_LIST_FORMAT_V1})

# On-wire Merkle list-commitment algorithm identifier. The `tree_alg` field of
# the leaves-list pins the tree's hash algorithm; v1 registers a single value.
DEFAULT_TREE_ALG: str = "rfc9162-sha256"


# === Exception types ==========================================================


class SchemaMerkleLeavesFormatUnsupported(Exception):
    """Raised by `decode_leaves_list` when `format` is not in the registered set."""

    code: str = "SCHEMA_MERKLE_LEAVES_FORMAT_UNSUPPORTED"

    def __init__(self, message: str = "") -> None:
        super().__init__(f"{self.code}: {message}" if message else self.code)


class MerkleLeavesListMalformed(Exception):
    """Raised when the decoded leaves-list violates a structural invariant
    (wrong types, missing required field, wrong-length root / leaf, etc.)."""

    code: str = "SCHEMA_MERKLE_LEAVES_MALFORMED"

    def __init__(self, message: str) -> None:
        super().__init__(f"{self.code}: {message}")


# === Decoded type (returned by `decode_leaves_list`) ==========================


@dataclass(frozen=True)
class LeavesList:
    format: str
    tree_alg: str
    root: bytes
    leaves: tuple[bytes, ...]
    leaf_count: int
    leaf_alg: str | None = None


# === CBOR encode / decode =====================================================


def encode_leaves_list(
    *,
    root: bytes,
    leaves: list[bytes] | tuple[bytes, ...],
    leaf_count: int,
    leaf_alg: str | None = None,
) -> bytes:
    """Emit canonical CBOR bytes for the leaves-list.

    Inputs:
      - `root`        — the 32-byte Merkle root (raw bytes, not hex).
      - `leaves`      — ordered list of 32-byte leaf digests.
      - `leaf_count`  — total leaf count (must equal `len(leaves)`).
      - `leaf_alg`    — OPTIONAL informative leaf-pre-hash algorithm.

    The output is canonical CBOR (RFC 8949 §4.2.1): shortest-form integers,
    definite-length encoding, bytewise lex-sorted map keys, no duplicates.
    """
    if not isinstance(root, (bytes, bytearray)) or len(root) != 32:
        raise MerkleLeavesListMalformed("root must be a 32-byte byte string")
    if not isinstance(leaves, (list, tuple)) or len(leaves) == 0:
        raise MerkleLeavesListMalformed("leaves must be a non-empty list")
    if leaf_count != len(leaves):
        raise MerkleLeavesListMalformed(
            f"leaf_count ({leaf_count}) does not match len(leaves) ({len(leaves)})"
        )
    for i, leaf in enumerate(leaves):
        if not isinstance(leaf, (bytes, bytearray)) or len(leaf) != 32:
            raise MerkleLeavesListMalformed(f"leaves[{i}] must be a 32-byte byte string")

    obj: dict[str, Any] = {
        "format": LEAVES_LIST_FORMAT_V1,
        "tree_alg": DEFAULT_TREE_ALG,
        "root": bytes(root),
        "leaves": [bytes(leaf) for leaf in leaves],
        "leaf_count": leaf_count,
    }
    if leaf_alg is not None:
        obj["leaf_alg"] = leaf_alg
    return encode_canonical_cbor(obj)


def decode_leaves_list(data: bytes) -> LeavesList:
    """Parse canonical CBOR bytes into a `LeavesList`.

    Raises:
      - `SchemaMerkleLeavesFormatUnsupported` — `format` not in registered set.
      - `MerkleLeavesListMalformed` — structural violation (wrong types,
        missing required field, wrong-length root / leaf, leaf_count
        disagreement with len(leaves), etc.).
      - `ValueError` from the canonical CBOR decoder for `MALFORMED_CBOR`
        upstream (floats, indefinite-length, etc.).
    """
    decoded = decode_canonical_cbor(data)
    if not isinstance(decoded, dict):
        raise MerkleLeavesListMalformed("leaves-list top-level must be a CBOR map")

    fmt = decoded.get("format")
    if not isinstance(fmt, str):
        raise MerkleLeavesListMalformed("leaves-list `format` must be a text string")
    if fmt not in REGISTERED_FORMATS:
        raise SchemaMerkleLeavesFormatUnsupported(f"unsupported leaves-list format: {fmt!r}")

    tree_alg = decoded.get("tree_alg")
    if not isinstance(tree_alg, str):
        raise MerkleLeavesListMalformed("leaves-list `tree_alg` must be a text string")

    root = decoded.get("root")
    if not isinstance(root, (bytes, bytearray)) or len(root) != 32:
        raise MerkleLeavesListMalformed("leaves-list `root` must be a 32-byte byte string")

    leaves_raw = decoded.get("leaves")
    if not isinstance(leaves_raw, list) or len(leaves_raw) == 0:
        raise MerkleLeavesListMalformed("leaves-list `leaves` must be a non-empty array")
    leaves: list[bytes] = []
    for i, leaf in enumerate(leaves_raw):
        if not isinstance(leaf, (bytes, bytearray)) or len(leaf) != 32:
            raise MerkleLeavesListMalformed(
                f"leaves-list `leaves[{i}]` must be a 32-byte byte string"
            )
        leaves.append(bytes(leaf))

    leaf_count = decoded.get("leaf_count")
    if not isinstance(leaf_count, int) or isinstance(leaf_count, bool) or leaf_count < 0:
        raise MerkleLeavesListMalformed("leaves-list `leaf_count` must be a non-negative integer")
    if leaf_count != len(leaves):
        raise MerkleLeavesListMalformed(
            f"leaves-list `leaf_count` ({leaf_count}) does not match len(leaves) ({len(leaves)})"
        )

    leaf_alg = decoded.get("leaf_alg")
    if leaf_alg is not None and not isinstance(leaf_alg, str):
        raise MerkleLeavesListMalformed("leaves-list `leaf_alg` (if present) must be a text string")

    return LeavesList(
        format=fmt,
        tree_alg=tree_alg,
        root=bytes(root),
        leaves=tuple(leaves),
        leaf_count=leaf_count,
        leaf_alg=leaf_alg,
    )


# === JSON projection (informative) ============================================


def to_json_projection(decoded: LeavesList) -> str:
    """Return the JCS-canonical (RFC 8785) JSON projection of `decoded`.

    The JSON projection encodes `root` and every `leaves[i]` as 64-character
    lowercase hex strings (no `0x` prefix). This is an INSPECTION-ONLY view;
    verification MUST be performed against the CBOR canonical form.
    """
    obj: dict[str, Any] = {
        "format": decoded.format,
        "tree_alg": decoded.tree_alg,
        "root": decoded.root.hex(),
        "leaves": [leaf.hex() for leaf in decoded.leaves],
        "leaf_count": decoded.leaf_count,
    }
    if decoded.leaf_alg is not None:
        obj["leaf_alg"] = decoded.leaf_alg
    # JCS canonical form (RFC 8785): lexicographic key order, no whitespace,
    # UTF-8 codepoints not escaped.
    return json.dumps(obj, sort_keys=True, ensure_ascii=False, separators=(",", ":"))


def from_json_projection(json_str: str) -> LeavesList:
    """Parse a JSON projection back into a `LeavesList`.

    WARNING: the JSON projection is NOT the wire form. Verifiers MUST parse
    CBOR (`decode_leaves_list`) for any check whose outcome affects record
    validity. This helper exists for CLI tooling and documentation round-trips
    only.

    Raises the same exception classes as `decode_leaves_list`.
    """
    try:
        obj = json.loads(json_str)
    except json.JSONDecodeError as e:
        raise MerkleLeavesListMalformed(f"leaves-list JSON parse failed: {e}") from e
    if not isinstance(obj, dict):
        raise MerkleLeavesListMalformed("leaves-list JSON top-level must be an object")

    fmt = obj.get("format")
    if not isinstance(fmt, str):
        raise MerkleLeavesListMalformed("leaves-list `format` must be a string")
    if fmt not in REGISTERED_FORMATS:
        raise SchemaMerkleLeavesFormatUnsupported(f"unsupported leaves-list format: {fmt!r}")

    tree_alg = obj.get("tree_alg")
    if not isinstance(tree_alg, str):
        raise MerkleLeavesListMalformed("leaves-list `tree_alg` must be a string")

    root_hex = obj.get("root")
    if not isinstance(root_hex, str) or len(root_hex) != 64:
        raise MerkleLeavesListMalformed("leaves-list JSON `root` must be a 64-character hex string")
    try:
        root = bytes.fromhex(root_hex)
    except ValueError as e:
        raise MerkleLeavesListMalformed(f"leaves-list `root` hex decode failed: {e}") from e

    leaves_raw = obj.get("leaves")
    if not isinstance(leaves_raw, list) or len(leaves_raw) == 0:
        raise MerkleLeavesListMalformed("leaves-list `leaves` must be a non-empty array")
    leaves: list[bytes] = []
    for i, leaf_hex in enumerate(leaves_raw):
        if not isinstance(leaf_hex, str) or len(leaf_hex) != 64:
            raise MerkleLeavesListMalformed(
                f"leaves-list JSON `leaves[{i}]` must be a 64-character hex string"
            )
        try:
            leaves.append(bytes.fromhex(leaf_hex))
        except ValueError as e:
            raise MerkleLeavesListMalformed(
                f"leaves-list `leaves[{i}]` hex decode failed: {e}"
            ) from e

    leaf_count = obj.get("leaf_count")
    if not isinstance(leaf_count, int) or isinstance(leaf_count, bool) or leaf_count < 0:
        raise MerkleLeavesListMalformed("leaves-list `leaf_count` must be a non-negative integer")
    if leaf_count != len(leaves):
        raise MerkleLeavesListMalformed(
            f"leaves-list `leaf_count` ({leaf_count}) does not match len(leaves) ({len(leaves)})"
        )

    leaf_alg = obj.get("leaf_alg")
    if leaf_alg is not None and not isinstance(leaf_alg, str):
        raise MerkleLeavesListMalformed("leaves-list `leaf_alg` (if present) must be a string")

    return LeavesList(
        format=fmt,
        tree_alg=tree_alg,
        root=root,
        leaves=tuple(leaves),
        leaf_count=leaf_count,
        leaf_alg=leaf_alg,
    )


__all__ = [
    "DEFAULT_TREE_ALG",
    "LEAVES_LIST_FORMAT_V1",
    "REGISTERED_FORMATS",
    "LeavesList",
    "MerkleLeavesListMalformed",
    "SchemaMerkleLeavesFormatUnsupported",
    "decode_leaves_list",
    "encode_leaves_list",
    "from_json_projection",
    "to_json_projection",
]
